File: 4_leetcode/C++/LC_1114_mut.cpp
#include <mutex>
#include <condition_variable>
#include <functional>

using namespace std;

class Foo {
private:
    mutex mtx;
    condition_variable cv;
    int count;

public:
    Foo() : count(0) {}

    void first(function<void()> printFirst) {
        unique_lock<mutex> lock(mtx);
        
        // printFirst() outputs "first". Do not change or remove this line.
        printFirst();
        
        count++;
        cv.notify_all();
    }

    void second(function<void()> printSecond) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [this]() { return count>=1; });
        //while(obj->x!=1) pthread_cond_wait(&obj->cond, &obj->mut);
        
        // printSecond() outputs "second". Do not change or remove this line.
        printSecond();
        
        count++;
        cv.notify_all();
    }

    void third(function<void()> printThird) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [this]() { return count>=2; });
        
        // printThird() outputs "third". Do not change or remove this line.
        printThird();
        
        count=0; 
    }
};

File: 4_leetcode/C#/LC_1114_mut.cs
using System;
using System.Threading;

public class Foo {
    
    private int x;
    private object mut;
    private bool cond;

    public Foo() {
        x=1;

        mut =new object(); 
        cond=false; 
    }

    public void First(Action printFirst) {
        lock (mut) {
            while (x!=1) Monitor.Wait(mut); // wait
            

            x++;
            // printFirst() outputs "first". Do not change or remove this line.
            printFirst(); 

            Monitor.PulseAll(mut); // broadcast for all waiting threads
        }
        
        
    }

    public void Second(Action printSecond) {
        lock (mut){
            while (x!=2) Monitor.Wait(mut); // wait
            

            x++;
            // printSecond() outputs "second". Do not change or remove this line.
            printSecond();
            Monitor.PulseAll(mut); // broadcast for all waiting threads
        }
        
    }

    public void Third(Action printThird) {
        lock (mut){
            while (x!=3) Monitor.Wait(mut); // wait
            

            // printThird() outputs "third". Do not change or remove this line.
            printThird();
            Monitor.PulseAll(mut); // broadcast for all waiting threads
        }
        
    }
}